Prevent CiliumEndpoint removal by non-owning agent

timoreimann opened the pull request:
CEPs are creating as well as updated based on informer store data local to an agent's node but (necessarily) deleted globally from the API server. This can currently lead to situations where an agent that does not own a CEP deletes an unrelated CEP.

Avoid this problem by having agents maintain the CEP UID and using it as a precondition when deleting CEPs. This guarantees that only the owning agents can delete "their" CEPs.

Fixes #17047

**Note:** I am not yet sure if/how to best extend any automated testing: most of the files I updated seem to have many tests, so I'm a bit uncertain if initial tests need to be built out first or if coverage is maintained through other flavors of tests. I know Cilium has e2e tests as well, but I couldn't spot the corresponding tests for the change in question either. Appreciate any pointers in this regard.

See also the related Slack convo [here](https://cilium.slack.com/archives/C53TG4J4R/p1644882661331769).

/cc @aanm 

```release-note
Fix issue where StatefulSet pod restarts could trigger persistent connectivity issues for the pods due to overzealous CiliumEndpoint resource removal by cilium-agent instances
```

timoreimann committed:
Prevent CiliumEndpoint removal by non-owning agent

CEPs are creating as well as updated based on informer store data local
to an agent's node but (necessarily) deleted globally from the API
server. This can currently lead to situations where an agent that does
not own a CEP deletes an unrelated CEP.

Avoid this problem by having agents maintain the CEP UID and using it as
a precondition when deleting CEPs. This guarantees that only the owning
agents can delete "their" CEPs.

Signed-off-by: Timo Reimann <ttr314@googlemail.com>

christarazi commented on pkg/k8s/slim/k8s/api/core/v1/types.go:
```diff
@@ -252,6 +252,11 @@ type PodSpec struct {
+	// the scheduler simply schedules this pod onto that node, assuming that it fits resource
+	// requirements.
+	// +optional
+	NodeName string `json:"nodeName,omitempty" protobuf:"bytes,10,opt,name=nodeName"`
```
There's no need for this field actually because we know that any Pod object that enters the K8s watchers is a local pod due to https://github.com/cilium/cilium/blob/82c742f2e9fb65fb3fc392c7dcb3b4e22b69c650/pkg/k8s/watchers/pod.go#L122.

timoreimann replied:
The addition was based on a suggestion from @aanm in a discussion we had on Slack (with the relevant portion starting around [here](https://cilium.slack.com/archives/C53TG4J4R/p1645045322460259?thread_ts=1644882661.331769&cid=C53TG4J4R)). The idea was to play it safe since some code paths to get the pod seem to rely on [`GetCachedPod()`](https://github.com/cilium/cilium/blob/fcbeb6402b2ce63be373e27a216c4b6806188def/pkg/k8s/watchers/pod.go#L852), which comes with a comment saying that a `K8sEventHandover` determines whether the backing store is populated from local or cluster-wide pods. So the concern was that we could be seeing cluster-wide pods if the Cilium configuration was set accordingly.

With this additional context, can you confirm that omitting the `NodeName` addition (and consumption) is still the right thing to do?

christarazi replied:
As discussed in Slack, we'll keep the nodeName addition for a future PR to fix an endpoint restoration bug (sep. from this issue). That issue is tracked in https://github.com/cilium/cilium/issues/18923.

christarazi commented on pkg/k8s/watchers/endpointsynchronizer.go:
```diff
@@ -281,6 +297,21 @@ func (epSync *EndpointSynchronizer) RunK8sCiliumEndpointSync(e *endpoint.Endpoin
+// updateCEPUIDIfNeeded updates the endpoint's CEP UID from the local CEP if the
+// node name from the pod matches the agent's node name and the CEP UID is
+// different (i.e., has never been set on the endpoint or has changed).
+func updateCEPUIDIfNeeded(scopedLog *logrus.Entry, pod *slim_corev1.Pod, e *endpoint.Endpoint, localCEP *cilium_v2.CiliumEndpoint) {
```
Do we need to think about whether we need to lock the `Endpoint` object? We are unconditionally writing into it.

timoreimann replied:
I found it hard enough to judge if we needed to synchronize access or not, so I went ahead and added a getter/setter pair.

christarazi replied:
Yes, it can be hard to ascertain just from reading the code. If we take a step back and think about how this code is executing by following the callers / callees of the relevant functions, we can notice that it's called from async points (driven by K8s events) which is enough to tell us that it's possible to have a race, and hence need a lock. Thanks for adding it.

christarazi commented on pkg/k8s/watchers/endpointsynchronizer.go:
```diff
@@ -323,8 +354,31 @@ func deleteCEP(ctx context.Context, scopedLog *logrus.Entry, ciliumClient v2.Cil
+	// through unique pod names.) Storing the CEP UID at CEP create/fetch time
+	// and using it as a precondition for deletion ensures that agents may only
+	// delete CEPs they own.
+	// It is possible for the CEP UID to not be populated when an agent tries to
+	// clean up a CEP. In that case, skip deletion and rely on cilium operator
+	// garbage collection to clean up eventually.
```
This makes sense to me. I took a look at our CEP GC code and it's not quite as "observable" as I'd like if we now more explicitly rely on the GC to get the cluster into a good state. Specifically, we only have a debug level log, which I'd argue should stay that way but we need to augment it.

Could you add a new commit which exposes a metric for how many times we GC a CEP? The code should be around here: https://github.com/cilium/cilium/blob/d783aca9aa277397d60361de0af96da5a4bcf4f0/operator/k8s_cep_gc.go#L149

timoreimann replied:
Took a stab at this. Let me know what you think.

christarazi commented on pkg/k8s/watchers/endpointsynchronizer.go:
```diff
@@ -176,12 +184,14 @@ func (epSync *EndpointSynchronizer) RunK8sCiliumEndpointSync(e *endpoint.Endpoin
 							return err
 						}
 
+						scopedLog.WithField("cep_uid", localCEP.UID).Debug("storing CEP UID after create")
```
Let's define `cepGUID` (nit on underscores just to conform to existing camel-case) as a logfield inside the logfields package and use it throughout. It seems useful enough now

christarazi commented on pkg/k8s/watchers/endpointsynchronizer.go:
```diff
@@ -281,6 +297,21 @@ func (epSync *EndpointSynchronizer) RunK8sCiliumEndpointSync(e *endpoint.Endpoin
+			"agent_node_name": types.GetName(),
+			"pod_node_name":   pod.Spec.NodeName,
+			"old_cep_uid":     e.CiliumEndpointUID,
+			"new_cep_uid":     localCEP.UID,
```
```
			logfields.Node: types.GetName(),
			"pod_node_name":   pod.Spec.NodeName,  // <- no need as per above comment
			"old"+logfields.CEPGUID:     e.CiliumEndpointUID,
			logfield.CEPGUID:     localCEP.UID,
```

timoreimann replied:
Done. Left the pod's node name in for now until we have resolved the other conversation.

christarazi reviewed (changes requested):
Thanks for the PR! :rocket: 

A couple of comments below. Also, please squash the second commit if it's meant to change the same files as the first commit.

timoreimann committed:
Add metrics for endpoint objects garbage collection

Signed-off-by: Timo Reimann <ttr314@googlemail.com>

christarazi commented on operator/k8s_cep_gc.go:
```diff
@@ -58,17 +60,38 @@ func enableCiliumEndpointSyncGC(once bool) {
+	cegc := CiliumEndpointGC{}
+	if operatorOption.Config.EnableMetrics {
+		cegc.objectsMetric = metrics.EndpointGCObjects
+	}
```
What's the need to do this?

timoreimann replied:
It is meant to encapsulate the logic that determines if metrics should be tracked or not based on the `EnableMetrics` flag: if the flag is enabled, `cegc.objectsMetric` will be populated and otherwise left nil. The convenience methods defined on the struct check `cegc.objectsMetric` before recording a successful/failure metric event.

The larger intention of mine was to not sprinkle `if` checks for whether metrics are enabled across the core business logic and instead scope it to the struct. This should also make it easier to test the logic as it does not rely on globals (though I haven't added tests yet).

Let me know if you think there's a different/better way to do it.

christarazi replied:
Ah I see, I think this can be simpler. I don't think we need a struct, but rather just change the helper functions to do the following

```go
func (cegc *CiliumEndpointGC) successfulObjectGC() {
	if metricsEnabled {
		metrics.EndpointGCObjects.WithLabelValues(metrics.LabelValueOutcomeSuccess).Inc()
	}
}
```

This does what you were intending, right? I don't think it's a deal breaker to have the if-statements checking the option, so we might as well simplify and remove the struct.

timoreimann replied:
Okay, done.

timoreimann committed:
Include nodeName on pods

Needed for a follow-up change to address an endpoint restoration issue
(https://github.com/cilium/cilium/issues/18923).

Signed-off-by: Timo Reimann <ttr314@googlemail.com>

christarazi reviewed (approved):
Thanks!
